crates/turbo-tasks/src/join_iter_ext.rs

```diff
@@ -67,24 +64,23 @@ where
 
 pub struct TryFlatMapRecursiveJoin<T, C, F, CI>
 where
-    T: Hash + PartialEq + Eq + Clone,
-    C: Fn(T) -> F,
+    C: FnMut(&T) -> Option<F>,
     F: Future<Output = Result<CI>>,
     CI: IntoIterator<Item = T>,
 {
-    set: IndexSet<T>,
-    futures: FuturesOrdered<F>,
-    get_children: C,
+    output: Vec<T>,
+    futures: FuturesUnordered<F>,
```

Comment:
```suggestion
    futures: FuturesUnordered<WithParent<T, F>>,
    roots: Vec<T>,
```

```
WithParent<T, F> = impl Future<Output = Result<(T, Vec<T>)>>
```

Tuple is `(parent, children)`

So you can keep the edges directly including order...